### PR TITLE
[Enhancement] Introduce varbinary type

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Like other database systems, `varbinary`  represents  a sequence of bytes with no paddings,  has no notion of Unicode characters which is different from `varchar` / `char`. In BE, we can just use `BinaryColumn` to store data of `varbinary` type, and its max length is same with `varchar` type.
- Add new varbinary type to distinguish the difference with `varchar` or `string`;
- Add VarbinaryLiteral to support binary inputs in SQL;
- Support hex(varbinary) function; 

Varbinary usage:
```
 create table t0(c0 INT, c1 varbinary(4), c2 varbinary, c3 varchar) duplicate key(c0) distributed by hash(c0) buckets 3 properties('replication_num'='1');
```
- varbinary(length): when max length is specific, exceptions will be thrown when  input bytes' length is greater than the specific length;
- varbinary: when max length is not set, its max length is default as `varchar`(https://docs.starrocks.io/zh-cn/latest/sql-reference/sql-statements/data-types/VARCHAR).

After this, we can do more things later:
- support `varbinary` type in distribution keys/ primary keys;
- support more `varbinary` functions. 

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 79 / 228 (34.65%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|0|2|00.00%|[159, 256]|
|:large_blue_circle:|com/starrocks/common/io/Text.java|0|8|00.00%|[397, 398, 399, 400, 414, 415, 416, 417]|
|:large_blue_circle:|com/starrocks/analysis/BinaryLiteral.java|0|68|00.00%|[45, 46, 47, 48, 53, 54, 55, 58, 59, 60, 61, 62, 64, 65, 66, 67, 68, 70, 71, 73, 74, 75, 78, 79, 80, 81, 85, 90, 91, 94, 95, 96, 97, 99, 101, 102, 103, 104, 105, 108, 109, 110, 112, 114, 115, 116, 118, 121, 127, 132, 137, 142, 143, 144, 148, 153, 158, 163, 168, 169, 170, 173, 174, 175, 178, 179, 180, 185]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|0|2|00.00%|[197, 198]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|2|00.00%|[580, 581]|
|:large_blue_circle:|com/starrocks/analysis/VarBinaryLiteral.java|16|71|22.54%|[54, 55, 56, 59, 60, 61, 62, 63, 79, 80, 81, 82, 86, 91, 92, 95, 96, 97, 98, 100, 102, 103, 104, 105, 106, 109, 110, 111, 113, 115, 116, 117, 119, 122, 128, 133, 143, 144, 145, 154, 159, 164, 169, 170, 171, 174, 175, 176, 179, 180, 181, 185, 186, 188, 193]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|11|16|68.75%|[92, 498, 499, 679, 706]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|4|5|80.00%|[124]|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|16|19|84.21%|[267, 495, 543]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|22|25|88.00%|[664, 1103, 1104]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Util.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by ZiheLiu** [comment]
We use `RunTimeTypeLimits<>::max_value` in some comparison logics, the `max_value` for binary type is `0xff`.
It is OK for VARCHAR, because `0xff` is an invalid UTF-8 VAHRCHAR.
Is it also OK for` varbinary`?
```C++
template <PrimitiveType ptype>
struct RunTimeTypeLimits<ptype, BinaryPTGuard<ptype>> {
    using value_type = RunTimeCppType<ptype>;

    static constexpr value_type min_value() { return Slice(&_min, 0); }
    static constexpr value_type max_value() { return Slice(&_max, 1); }

private:
    static inline char _min = 0x00;
    static inline char _max = 0xff;
};
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 81 / 162 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|0|2|00.00%|[159, 256]|
|:large_blue_circle:|com/starrocks/common/io/Text.java|0|8|00.00%|[397, 398, 399, 400, 414, 415, 416, 417]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|0|2|00.00%|[197, 198]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|2|00.00%|[580, 581]|
|:large_blue_circle:|com/starrocks/analysis/VarBinaryLiteral.java|16|71|22.54%|[35, 36, 37, 40, 41, 42, 43, 44, 60, 61, 62, 63, 67, 72, 73, 76, 77, 78, 79, 81, 83, 84, 85, 86, 87, 90, 91, 92, 94, 96, 97, 98, 100, 103, 109, 114, 124, 125, 126, 135, 140, 145, 150, 151, 152, 155, 156, 157, 160, 161, 162, 166, 167, 169, 174]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|11|16|68.75%|[92, 498, 499, 679, 706]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|4|5|80.00%|[124]|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|16|19|84.21%|[267, 495, 543]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|22|25|88.00%|[664, 1103, 1104]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Util.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 81 / 161 (50.31%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|0|2|00.00%|[159, 256]|
|:large_blue_circle:|com/starrocks/common/io/Text.java|0|8|00.00%|[397, 398, 399, 400, 414, 415, 416, 417]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|0|2|00.00%|[197, 198]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|1|00.00%|[580]|
|:large_blue_circle:|com/starrocks/analysis/VarBinaryLiteral.java|16|71|22.54%|[35, 36, 37, 40, 41, 42, 43, 44, 60, 61, 62, 63, 67, 72, 73, 76, 77, 78, 79, 81, 83, 84, 85, 86, 87, 90, 91, 92, 94, 96, 97, 98, 100, 103, 109, 114, 124, 125, 126, 135, 140, 145, 150, 151, 152, 155, 156, 157, 160, 161, 162, 166, 167, 169, 174]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|11|16|68.75%|[92, 498, 499, 679, 706]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|4|5|80.00%|[124]|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|16|19|84.21%|[267, 495, 543]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|22|25|88.00%|[664, 1103, 1104]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Util.java|1|1|100.00%|[]|

**Comment by LiShuMing** [comment]
> ```c++
> RunTimeTypeLimits
> ```

It's a problem, binary column(not char/varchar) should not have a min/max. I'll fix this.

**Review by murphyatwork — COMMENTED** [review]
Is network serialization/deserialization not implemented ?

File: be/src/runtime/primitive_type.h [hunk 1]
```diff
@@ -63,10 +63,21 @@ enum PrimitiveType {
...
-    TYPE_JSON,     /* 27 */
-    TYPE_FUNCTION, /* 28 */
+    TYPE_JSON,      /* 27 */
+    TYPE_FUNCTION,  /* 28 */
+    TYPE_VARBINARY, /* 29 */
 };
 
+inline bool is_binary_type(PrimitiveType type) {
+    switch (type) {
+    case TYPE_BINARY:
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply be/src/runtime/primitive_type.h:73]: `Binary` like the `Char` which will pad zero bits in `mysql` however varbinary is like varchar which will not add padding spaces.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 81 / 162 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|0|2|00.00%|[159, 256]|
|:large_blue_circle:|com/starrocks/common/io/Text.java|0|8|00.00%|[397, 398, 399, 400, 414, 415, 416, 417]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|0|2|00.00%|[197, 198]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|2|00.00%|[579, 580]|
|:large_blue_circle:|com/starrocks/analysis/VarBinaryLiteral.java|16|71|22.54%|[35, 36, 37, 40, 41, 42, 43, 44, 60, 61, 62, 63, 67, 72, 73, 76, 77, 78, 79, 81, 83, 84, 85, 86, 87, 90, 91, 92, 94, 96, 97, 98, 100, 103, 109, 114, 124, 125, 126, 135, 140, 145, 150, 151, 152, 155, 156, 157, 160, 161, 162, 166, 167, 169, 174]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|11|16|68.75%|[92, 498, 499, 679, 706]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|4|5|80.00%|[124]|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|16|19|84.21%|[267, 495, 543]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|22|25|88.00%|[666, 1105, 1106]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Util.java|1|1|100.00%|[]|

**Comment by LiShuMing** [comment]
run starrocks_be_unittest

**Comment by LiShuMing** [comment]
> 

BinaryType now don't support min/max for now. I have changed original min/max type as StringType.

**Comment by LiShuMing** [comment]
run starrocks_admit_test

File: be/src/util/symbols_util.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+// e.g. for the literal ns1::ns2::class::type,
+// the dictionary would add 4 literals: 'ns1', 'ns1::ns2', 'ns1::ns2::class',
+//    'ns1::ns2::class::type'
+// We instead take some shortcuts since we know all the argument types are
+// types we define.
+
+// Mangled symbols must start with this.
+const char* MANGLE_PREFIX = "_Z";
+
+bool SymbolsUtil::is_mangled(const std::string& symbol) {
```
Comment (kangkaisen) [inline be/src/util/symbols_util.cpp:59]: why need this class?

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 81 / 162 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|0|2|00.00%|[160, 257]|
|:large_blue_circle:|com/starrocks/common/io/Text.java|0|8|00.00%|[397, 398, 399, 400, 414, 415, 416, 417]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|0|2|00.00%|[197, 198]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|2|00.00%|[585, 586]|
|:large_blue_circle:|com/starrocks/analysis/VarBinaryLiteral.java|16|71|22.54%|[35, 36, 37, 40, 41, 42, 43, 44, 60, 61, 62, 63, 67, 72, 73, 76, 77, 78, 79, 81, 83, 84, 85, 86, 87, 90, 91, 92, 94, 96, 97, 98, 100, 103, 109, 114, 124, 125, 126, 135, 140, 145, 150, 151, 152, 155, 156, 157, 160, 161, 162, 166, 167, 169, 174]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|11|16|68.75%|[92, 505, 506, 686, 713]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|4|5|80.00%|[124]|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|16|19|84.21%|[267, 495, 543]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|22|25|88.00%|[666, 1105, 1106]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Util.java|1|1|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 72 / 153 (47.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|0|2|00.00%|[160, 257]|
|:large_blue_circle:|com/starrocks/common/io/Text.java|0|8|00.00%|[397, 398, 399, 400, 414, 415, 416, 417]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|0|2|00.00%|[197, 198]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|2|00.00%|[585, 586]|
|:large_blue_circle:|com/starrocks/analysis/VarBinaryLiteral.java|16|71|22.54%|[35, 36, 37, 40, 41, 42, 43, 44, 60, 61, 62, 63, 67, 72, 73, 76, 77, 78, 79, 81, 83, 84, 85, 86, 87, 90, 91, 92, 94, 96, 97, 98, 100, 103, 109, 114, 124, 125, 126, 135, 140, 145, 150, 151, 152, 155, 156, 157, 160, 161, 162, 166, 167, 169, 174]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|11|16|68.75%|[92, 505, 506, 686, 713]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|4|5|80.00%|[124]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|13|16|81.25%|[665, 1105, 1106]|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|16|19|84.21%|[267, 495, 543]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Util.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 81 / 162 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|0|2|00.00%|[160, 257]|
|:large_blue_circle:|com/starrocks/common/io/Text.java|0|8|00.00%|[397, 398, 399, 400, 414, 415, 416, 417]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|0|2|00.00%|[204, 205]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|2|00.00%|[620, 621]|
|:large_blue_circle:|com/starrocks/analysis/VarBinaryLiteral.java|16|71|22.54%|[35, 36, 37, 40, 41, 42, 43, 44, 60, 61, 62, 63, 67, 72, 73, 76, 77, 78, 79, 81, 83, 84, 85, 86, 87, 90, 91, 92, 94, 96, 97, 98, 100, 103, 109, 114, 124, 125, 126, 135, 140, 145, 150, 151, 152, 155, 156, 157, 160, 161, 162, 166, 167, 169, 174]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|11|16|68.75%|[92, 505, 506, 686, 713]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|4|5|80.00%|[130]|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|16|19|84.21%|[267, 495, 543]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|22|25|88.00%|[690, 1129, 1130]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Util.java|1|1|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**